OneCommunityGlobal/HGNRest
# Nathan fix dev admin put permissions

# Description
This is a back-end implementation of [Frontend PR 994](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/994)

## Related PRS (if any):
This was branched off of PR #416.

## Main changes explained:
Adds ``canRequestorUpdateUser()`` which just checks if the user being modified is the devAdmin account and if the requestor is the only account that should be able to modify it. That function is then called in ``putUserProfile()`` and ``updatepassword()`` to prevent any admin accounts from editing the devAdmin account.


## How to test:
1. check into current branch
2. do `npm install` and `...` to run this PR locally
3. Open postman
4. Post login with admin credentials. Returns a token.
5. Copy the returned token into the Authorization header for PUT Profiles_PutById
6. Make sure PUT Profiles_PutById is for the devAdmin account ``64c17eb8c737b05dd4ac4e28``. The link should be: http://localhost:4500/api/userprofile/64c17eb8c737b05dd4ac4e28
7. Copy the below payload into the body. This would add devAdmin to the "Joyce Project" project.
8. Send the request.
9. The response should be 403 Forbidden with the body "You are not authorized to update this user"
10. You should also be able to confirm in [the app](https://hgnapplication_react_dev.surge.sh/userprofile/64c17eb8c737b05dd4ac4e28) that DevAdmin was not added to the "Joyce Project" project.

## Screenshots or videos of changes:
![image](https://github.com/OneCommunityGlobal/HGNRest/assets/23086054/3c58eaeb-1a89-4627-bfd6-a1ef60f6cdf0)
![image](https://github.com/OneCommunityGlobal/HGNRest/assets/23086054/de9b7ffb-40c7-4101-818e-db5de08e9d18)


## Note:
[Postman commands](https://documenter.getpostman.com/view/5808634/RztfxY4M#151f24e9-1b65-48e7-b2d7-240eea118333)

Payload:
```
{
    "permissions": {
        "frontPermissions": [],
        "backPermissions": []
    },
    "privacySettings": {
        "blueSquares": true,
        "email": true,
        "phoneNumber": true
    },
    "hoursByCategory": {
        "housing": 0,
        "food": 0,
        "education": 0,
        "society": 0,
        "energy": 0,
        "economics": 0,
        "stewardship": 0,
        "unassigned": 0
    },
    "isActive": true,
    "isRehireable": false,
    "isSet": false,
    "phoneNumber": [
        "18888888888"
    ],
    "jobTitle": [
        ""
    ],
    "weeklycommittedHours": 0,
    "missedHours": 0,
    "createdDate": "2023-07-26T07:00:00.000Z",
    "teams": [],
    "projects": [
        {
            "category": "Food",
            "_id": "64b67363031aa2c3841d136f",
            "projectName": "Joyce Project"
        },
        {
            "category": "Society",
            "_id": "6157b9a66dfe943665c9f83b",
            "projectName": "Orientation and Initial Setup"
        }
    ],
    "location": "",
    "weeklySummariesCount": 0,
    "personalBestMaxHrs": 0,
    "totalTangibleHrs": 0,
    "totalIntangibleHrs": 0,
    "lastWeekTangibleHrs": 0,
    "savedTangibleHrs": [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0
    ],
    "weeklySummaryNotReq": false,
    "timeZone": "America/Los_Angeles",
    "isVisible": false,
    "bioPosted": "default",
    "_id": "64c17eb8c737b05dd4ac4e28",
    "personalLinks": [],
    "adminLinks": [],
    "badgeCollection": [],
    "infringements": [],
    "oldInfringements": [],
    "weeklySummaries": [
        {
            "dueDate": "2023-08-06T06:59:59.999Z",
            "_id": "64c6a3547a22a959a8cb0f11",
            "summary": ""
        },
        {
            "dueDate": "2023-08-06T06:59:59.999Z",
            "_id": "64c68e56f1954450f8cf31b8",
            "summary": ""
        },
        {
            "dueDate": "2023-08-06T06:59:59.999Z",
            "_id": "64c60cfcff7eaee0754cf1e3",
            "summary": ""
        },
        {
            "dueDate": "2023-08-06T06:59:59.999Z",
            "_id": "64c60c3b5856cb7a6a3028c0",
            "summary": ""
        }
    ],
    "categoryTangibleHrs": [
        {
            "category": "Other",
            "hrs": 0,
            "_id": "64c2a20c91220a8e08fc50a7"
        }
    ],
    "timeEntryEditHistory": [],
    "role": "Administrator",
    "firstName": "DevAdmin",
    "lastName": "CreateYourOwnAccount",
    "bio": "",
    "email": "devadmin@hgn.net",
    "weeklySummaryOption": "Not Required",
    "mediaUrl": "",
    "collaborationPreference": "Zoom",
    "weeklycommittedHoursHistory": [
        {
            "_id": "64c327e31e7649fc1ec7d7df",
            "hours": 10,
            "dateChanged": "2023-07-26T20:14:48.637Z"
        }
    ],
    "tangibleHoursReportedThisWeek": "0.00"
}
```


In case you don't swap branches or another error occurs, you can use this payload to revert the account back:
```
{
    "permissions": {
        "frontPermissions": [],
        "backPermissions": []
    },
    "privacySettings": {
        "blueSquares": true,
        "email": true,
        "phoneNumber": true
    },
    "hoursByCategory": {
        "housing": 0,
        "food": 0,
        "education": 0,
        "society": 0,
        "energy": 0,
        "economics": 0,
        "stewardship": 0,
        "unassigned": 0
    },
    "isActive": true,
    "isRehireable": false,
    "isSet": false,
    "phoneNumber": [
        "18888888888"
    ],
    "jobTitle": [
        ""
    ],
    "weeklycommittedHours": 0,
    "missedHours": 0,
    "createdDate": "2023-07-26T07:00:00.000Z",
    "teams": [],
    "projects": [
        {
            "category": "Society",
            "_id": "6157b9a66dfe943665c9f83b",
            "projectName": "Orientation and Initial Setup"
        }
    ],
    "location": "",
    "weeklySummariesCount": 0,
    "personalBestMaxHrs": 0,
    "totalTangibleHrs": 0,
    "totalIntangibleHrs": 0,
    "lastWeekTangibleHrs": 0,
    "savedTangibleHrs": [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0
    ],
    "weeklySummaryNotReq": false,
    "timeZone": "America/Los_Angeles",
    "isVisible": false,
    "bioPosted": "default",
    "_id": "64c17eb8c737b05dd4ac4e28",
    "personalLinks": [],
    "adminLinks": [],
    "badgeCollection": [],
    "infringements": [],
    "oldInfringements": [],
    "weeklySummaries": [
        {
            "dueDate": "2023-08-06T06:59:59.999Z",
            "_id": "64c6a3547a22a959a8cb0f11",
            "summary": ""
        },
        {
            "dueDate": "2023-08-06T06:59:59.999Z",
            "_id": "64c68e56f1954450f8cf31b8",
            "summary": ""
        },
        {
            "dueDate": "2023-08-06T06:59:59.999Z",
            "_id": "64c60cfcff7eaee0754cf1e3",
            "summary": ""
        },
        {
            "dueDate": "2023-08-06T06:59:59.999Z",
            "_id": "64c60c3b5856cb7a6a3028c0",
            "summary": ""
        }
    ],
    "categoryTangibleHrs": [
        {
            "category": "Other",
            "hrs": 0,
            "_id": "64c2a20c91220a8e08fc50a7"
        }
    ],
    "timeEntryEditHistory": [],
    "role": "Administrator",
    "firstName": "DevAdmin",
    "lastName": "CreateYourOwnAccount",
    "bio": "",
    "email": "devadmin@hgn.net",
    "weeklySummaryOption": "Not Required",
    "mediaUrl": "",
    "collaborationPreference": "Zoom",
    "weeklycommittedHoursHistory": [
        {
            "_id": "64c327e31e7649fc1ec7d7df",
            "hours": 10,
            "dateChanged": "2023-07-26T20:14:48.637Z"
        }
    ],
    "tangibleHoursReportedThisWeek": "0.00"
}
```
